src/xenia/base/memory.cc

```diff
@@ -215,7 +219,102 @@ void copy_and_swap_16_in_32_unaligned(void* dest_ptr, const void* src_ptr,
+                              size_t count) {
+  copy_and_swap_16_unaligned(dst_ptr, src_ptr, count);
+}
+
+void copy_and_swap_16_unaligned(void* dst_ptr, const void* src_ptr,
+                                size_t count) {
+  auto dst = reinterpret_cast<uint16_t*>(dst_ptr);
+  auto src = reinterpret_cast<const uint16_t*>(src_ptr);
+
+  const uint8x16_t tbl_idx = vcombine_u8(vcreate_u8(0x0607040502030001),
+                                         vcreate_u8(0x0E0F0C0D0A0B0809));
+
+  int64_t i;
+  for (i = 0; i <= static_cast<int64_t>(count) - 8; i += 8) {
+    uint8x16_t data = vld1q_u8(reinterpret_cast<const uint8_t*>(&src[i]));
+    data = vqtbl1q_u8(data, tbl_idx);
+    vst1q_u8(reinterpret_cast<uint8_t*>(&dst[i]), data);
+  }
+
+  for (; i < count; ++i) {
+    dst[i] = byte_swap(src[i]);
+  }
+}
+
+void copy_and_swap_32_aligned(void* dst, const void* src, size_t count) {
+  copy_and_swap_32_unaligned(dst, src, count);
+}
+
+void copy_and_swap_32_unaligned(void* dst_ptr, const void* src_ptr,
+                                size_t count) {
+  auto dst = reinterpret_cast<uint32_t*>(dst_ptr);
+  auto src = reinterpret_cast<const uint32_t*>(src_ptr);
+
+  const uint8x16_t tbl_idx = vcombine_u8(vcreate_u8(0x405060700010203),
+                                         vcreate_u8(0x0C0D0E0F08090A0B));
+
+  int64_t i;
+  for (i = 0; i <= static_cast<int64_t>(count) - 4; i += 4) {
+    uint8x16_t data = vld1q_u8(reinterpret_cast<const uint8_t*>(&src[i]));
+    data = vqtbl1q_u8(data, tbl_idx);
+    vst1q_u8(reinterpret_cast<uint8_t*>(&dst[i]), data);
+  }
+
+  for (; i < count; ++i) {
+    dst[i] = byte_swap(src[i]);
+  }
+}
+
+void copy_and_swap_64_aligned(void* dst, const void* src, size_t count) {
+  copy_and_swap_64_unaligned(dst, src, count);
+}
+
+void copy_and_swap_64_unaligned(void* dst_ptr, const void* src_ptr,
+                                size_t count) {
+  auto dst = reinterpret_cast<uint64_t*>(dst_ptr);
+  auto src = reinterpret_cast<const uint64_t*>(src_ptr);
+
+  const uint8x16_t tbl_idx = vcombine_u8(vcreate_u8(0x0001020304050607),
+                                         vcreate_u8(0x08090A0B0C0D0E0F));
+
+  int64_t i;
+  for (i = 0; i <= static_cast<int64_t>(count) - 2; i += 2) {
+    uint8x16_t data = vld1q_u8(reinterpret_cast<const uint8_t*>(&src[i]));
+    data = vqtbl1q_u8(data, tbl_idx);
+    vst1q_u8(reinterpret_cast<uint8_t*>(&dst[i]), data);
+  }
+
+  for (; i < count; ++i) {
+    dst[i] = byte_swap(src[i]);
+  }
+}
+
+void copy_and_swap_16_in_32_aligned(void* dst, const void* src, size_t count) {
+  return copy_and_swap_16_in_32_unaligned(dst, src, count);
+}
+
+void copy_and_swap_16_in_32_unaligned(void* dst_ptr, const void* src_ptr,
+                                      size_t count) {
+  auto dst = reinterpret_cast<uint64_t*>(dst_ptr);
+  auto src = reinterpret_cast<const uint64_t*>(src_ptr);
+  for (size_t i = 0; i < count; ++i) {
+    dst[i] = (src[i] >> 16) | (src[i] << 16);
```

Comment:
Is vtbl more or less like pshufb, or is it not enough for 128-bit 16-in-32 swapping?

Reply:
vtbl is like shuffle, yes. I didn't bother to implement it since it's only used in vulkan code in one place afaik.
Also since it's not used in D3D12 right now, I figured it might not be needed anymore eventually.

Also if its 16 in 32 then why are the `dst` and `src` pointers of type `uint64_t*`??? I guess @bwrsandman stumbled across this as well and hence didn't add tests for it.

Reply:
Ah, yes, if I recall correctly, the gpu::PrimitiveProcessor has its own byte swapping logic combined with index buffer modifications. The vtbl code needs to be written there eventually as well, currently it uses vrev (if there's swapping code in it already at all). But in the main use case — GPUs without 32-bit vertex index support, in cases when manual index fetch cannot be done (such as with primitive restart) — raw copy and swap is not very useful anyway as the primitive restart index needs to be replaced with 0xFFFFFFFF as well while copying.

Regarding uint64_t — all I can say here probably is 🤔🤔🤔🙃🤷‍♂️

Reply:
It should be seriously broken. garbling half of the values (in 32bit units) and writing over the buffer end